javascript/ql/src/semmle/javascript/security/dataflow/ExternalAPIUsedWithUntrustedDataCustomizations.qll

```diff
@@ -0,0 +1,363 @@
+    // base case: import
+    exists(string path |
+      f = API::moduleImport(path) and
+      not path instanceof SafeExternalAPIPackage and
+      // Exclude paths that can be resolved to a file in the project
+      not exists(Import imprt |
+        imprt.getImportedPath().getValue() = path and exists(imprt.getImportedModule())
+      )
+    )
+    or
+    // covariant recursive cases: instances, members, results, and promise contents
+    // of something that comes from a library may themselves come from that library
+    exists(API::Node base | mayComeFromLibrary(base) |
+      f = base.getInstance() or
+      f = base.getAMember() or
+      f = base.getReturn() or
+      f = base.getPromised()
+    )
+    or
+    // contravariant recursive case: parameters of something that escapes to a library
+    // may come from that library
+    exists(API::Node base | mayEscapeToLibrary(base) | f = base.getAParameter())
+  }
+
+  /**
+   * Holds if data written to a definition of `f` may flow to an imported package.
+   */
+  private predicate mayEscapeToLibrary(API::Node f) {
+    // covariant recursive case: members, results, and promise contents of something that
+    // escapes to a library may themselves escape to that library
+    exists(API::Node base | mayEscapeToLibrary(base) and not isDeepObjectSink(base) |
+      f = base.getAMember() or
+      f = base.getPromised() or
+      f = base.getReturn()
+    )
+    or
+    // contravariant recursive case: arguments (other than the receiver) passed to a function
+    // that comes from a library may escape to that library
+    exists(API::Node base | mayComeFromLibrary(base) |
+      f = base.getAParameter() and not f = base.getReceiver()
+    )
+  }
+
+  /**
+   * Holds `node` may be part of an access path leading to an external API call.
+   */
+  private predicate nodeIsRelevant(API::Node node) {
+    mayComeFromLibrary(node) and
+    not node instanceof SafeExternalAPIFunction
+    or
+    nodeIsRelevant(node.getASuccessor()) and
+    not node = API::moduleImport(any(SafeExternalAPIPackage p))
+  }
+
+  /** Holds if the edge `pred -> succ` may lead to an external API call. */
+  private predicate edge(API::Node pred, API::Node succ) {
+    nodeIsRelevant(succ) and
+    pred.getASuccessor() = succ
+  }
+
+  /**
+   * Gets the depth of `node` from the API graph root, not including paths that go through
+   * irrelevant nodes, such as a package marked as safe.
+   */
+  private int getDepth(API::Node node) = shortestDistances(API::root/0, edge/2)(_, node, result)
+
+  /**
+   * Gets a parameter of `base` with name `name`, or a property named `name` of a destructuring parameter.
+   */
+  private API::Node getNamedParameter(API::Node base, string name) {
+    exists(API::Node param, string lbl |
+      // getParameter(i) requires a bindingset for i, so use the raw label
+      param = base.getASuccessor("parameter " + lbl) and
+      lbl != "-1" // ignore receiver
```

Comment:
Is this the same as `param = base.getAParameter() and not param = base.getReceiver()`, which you've also used above? If so, should we have a predicate for that?

Reply:
Changed to `param = base.getAParameter() and not param = base.getReceiver()` for now. Looks like I needed the parameter index at some earlier version of the predicate and forgot to clean it up.

It might make sense to look at the API for this. Generally I'd be in favor of being consistent with the data-flow API here (where the receiver isn't considered to be an argument or parameter) and just introduce the verbose `getParameterOrReceiver(i)` and `getAParameterOrReceiver` for the cases where you want both. We could introduce similar predicates in the data-flow API.

Reply:
That makes sense.

Reply:
Maybe in another PR. The refactoring would probably warrant more evaluations than I'd care to spend on this query.